Miscellaneous compiler/core refactorings

petebacondarwin opened the pull request:
See commits...

petebacondarwin committed:
refactor(compiler): remove unused function

The `accessExportScope()` function is not used anymore.

petebacondarwin committed:
refactor(compiler): remove unused `emitAllPartialModules()` and associated code

This method does not appear to be used in the project.
This commit removes it and code that it exclusively
depended upon, or depended upon it.

petebacondarwin committed:
refactor(core): use `unknown` rather than `never` for private properties

Before `unknown` was available, the `never` type was used to discourage
application developers from using "private" properties. The `unknown` type
is much better suited for this.

petebacondarwin committed:
refactor(core): remove `getFactoryOf` from `angularCoreDiEnv`

This function is never actually generated directly so it can be removed
from the context passed to the JIT compiler.

petebacondarwin commented on packages/core/src/render3/definition.ts:
```diff
@@ -288,67 +288,65 @@ export function ɵɵdefineComponent<T>(componentDefinition: {
    * The set of schemas that declare elements to be allowed in the component's template.
    */
   schemas?: SchemaMetadata[] | null;
-}): never {
```
Most of the changes in this file are just whitespace indentation.

petebacondarwin committed:
refactor(core): do not publish `ɵɵgetFactoryOf()`

Previously, `ɵɵgetFactoryOf()` was "privately" published from
`@angular/core` since in the past it was assumed that this
might be an instruction generated by the compiler.

This is not currently the case, so this commit removes it from
the private exports and renames it to indicate that it is a local
helper function.

JoostK commented on packages/compiler/src/aot/compiler.ts:
```diff
@@ -15,23 +15,16 @@ import {createTokenForExternalReference, Identifiers} from '../identifiers';
-import {compileNgModuleFromRender2 as compileR3Module} from '../render3/r3_module_compiler';
-import {compilePipeFromRender2 as compileR3Pipe} from '../render3/r3_pipe_compiler';
-import {htmlAstToRender3Ast} from '../render3/r3_template_transform';
-import {compileComponentFromRender2 as compileR3Component, compileDirectiveFromRender2 as compileR3Directive} from '../render3/view/compiler';
```
If you are doing this cleanup, there's a lot of logic in `render3/view/compiler` that should have gone unused now that this is deleted. all `*FromRender2` and `*FromGlobalMetadata` should be obsolete.

JoostK commented on packages/compiler/src/aot/compiler.ts:
```diff
@@ -15,23 +15,16 @@ import {createTokenForExternalReference, Identifiers} from '../identifiers';
 import * as o from '../output/output_ast';
 import {ParseError} from '../parse_util';
-import {compileNgModuleFromRender2 as compileR3Module} from '../render3/r3_module_compiler';
-import {compilePipeFromRender2 as compileR3Pipe} from '../render3/r3_pipe_compiler';
```
The render2 pipe compilation should also be dead code now, so can be removed.

petebacondarwin replied:
Gosh - that is exciting. I'll see if I can remove that stuff too...

JoostK commented on packages/core/test/render3/jit_environment_spec.ts:
```diff
@@ -18,6 +18,7 @@ const INTERFACE_EXCEPTIONS = new Set<string>([
   'ɵɵNgModuleDefWithMeta',
   'ɵɵPipeDefWithMeta',
   'ɵɵFactoryDef',
+  'ɵɵgetFactoryOf',
```
Instead of excluding here, could we go even further and actually remove it from `Identifiers` in `@angular/compiler` and remove the place where that is still referenced? (supposedly that is generation logic is dead code)

petebacondarwin replied:
I thought about that but I don't remember why I didn't do it. Perhaps I was just too scared. I'll give it a go. Thanks.

petebacondarwin replied:
This is why I didn't do it: https://github.com/angular/angular/blob/6f6102d8ad6627fd65133d1686e0e637a3f9990b/packages/compiler/src/render3/r3_factory.ts#L244

JoostK replied:
Yeah that's the one that I believe is dead; I can't find how `meta.delegateType === R3FactoryDelegateType.Factory` is ever true, given that `R3FactoryDelegateType.Factory` is never used as a value anywhere.

JoostK replied:
If this is not dead, then it shouldn't be deleted from the JIT environment.

petebacondarwin replied:
But strangely given that this PR actually removes that function, but doesn't fail, then either this code is redundant and never actually used or our test coverage is too low...

petebacondarwin replied:
After offline discussion with @JoostK this block and associated code can be removed too 😱 😱 😱 😱

JoostK reviewed (commented):
Type in the last commit:

> Previously, `ɵɵgetFactoryOf()` was "privately" published from
`@angular/core` since in the **passed** -> **past**

petebacondarwin committed:
fixup! refactor(core): remove `getFactoryOf` from `angularCoreDiEnv`

petebacondarwin committed:
fixup! refactor(compiler): remove unused `emitAllPartialModules()` and associated code

petebacondarwin committed:
refactor(compiler): remove unused `__BUILD_OPTIMIZER_*` constants

These constants were created in a very early phase of Ivy development.
They have never been used in the framework, no the build-optimizer tool.

AndrewKushnir commented:
[Presubmit](http://test/OCL:360680830:BASE:360680849:1614790417770:b5a4d935).

mhevery reviewed (approved):
reviewed-for: global-approvers
